### ISSUE TITLE
Intermittently receiving IndexOutOfBoundsException from AsyncDisposableTypeCache.IsAsyncDisposable in 5.2.0

### ISSUE THREAD

**Issue by kaevne** [simpleinjector/SimpleInjector#883]
We upgraded from 4.9.0 to 5.2.0 and from net45 to netstandard2.0, and now we're intermittently getting an IndexOutOfBoundsException from AsyncDisposableTypeCache.IsAsyncDisposable during container initialization randomly from a small number of servers.  All servers receive the same bits.  This is preventing process startup.

This is not happening in test environments and we can't seem to repro. 

One possibility is that we have Container EnableAutoVerification turned off to retain 4.9.0 behavior, so it's possible that multiple requests coming in at the same time during process startup and this new code is not thread-safe.  That being said, we had no issues on 4.9.0.

From the Windows Event logs, I've redacted some things for confidentiality:

> Event code: 3005 Event message: An unhandled exception has occurred. Event time: 1/12/2021 8:13:01 PM Event time (UTC): 1/12/2021 8:13:01 PM Event ID: 1f79f21e19124ce09b31568a4e9444fc 
Event sequence: 215782 Event occurrence: 215781 Event detail code: 0  Application information:     Application domain: /LM/W3SVC/2/ROOT/API-1-132549530037399294     Trust level: Full   
Application Virtual Path: /API     Application Path: C:\Program Files\-redacted-    Machine name: -redacted- Process information:     
Process ID: 138232     Process name: w3wp.exe     Account name: NT AUTHORITY\SYSTEM  Exception information:     Exception type: IndexOutOfRangeException     Exception message: Index 
was outside the bounds of the array.

```
at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
at SimpleInjector.Scope.AsyncDisposableTypeCache.IsAsyncDisposable(Object instance)
at SimpleInjector.Scope.CreateAndCacheInstanceInternal(ScopedRegistration registration, Dictionary`2 cache)
at SimpleInjector.Scope.GetInstanceInternal(ScopedRegistration registration)
at SimpleInjector.Scope.GetInstance[TImplementation](ScopedRegistration registration, Scope scope)
at SimpleInjector.Advanced.Internal.LazyScopedRegistration`1.GetInstance(Scope scope)
at lambda_method(Closure )
at SimpleInjector.Scope.CreateAndCacheInstanceInternal(ScopedRegistration registration, Dictionary`2 cache)
at SimpleInjector.Scope.GetInstanceInternal(ScopedRegistration registration)
at SimpleInjector.Scope.GetInstance[TImplementation](ScopedRegistration registration, Scope scope)
at SimpleInjector.Advanced.Internal.LazyScopedRegistration`1.GetInstance(Scope scope)
at lambda_method(Closure )
at SimpleInjector.InstanceProducer.GetInstance()
````

> Request information:     Request URL: https://-redacted-/api/v2.0/me/$batch     Request path: /api/v2.0/me/$batch     User host address: -redacted-     User: 
-redacted-    Is authenticated: True     Authentication Type: Bearer     Thread account name: NT AUTHORITY\SYSTEM  Thread information:     
Thread ID: 207     Thread account name: NT AUTHORITY\SYSTEM     Is impersonating: False

```
at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
at SimpleInjector.Scope.AsyncDisposableTypeCache.IsAsyncDisposable(Object instance)
at SimpleInjector.Scope.CreateAndCacheInstanceInternal(ScopedRegistration registration, Dictionary`2 cache)
at SimpleInjector.Scope.GetInstanceInternal(ScopedRegistration registration)
at SimpleInjector.Scope.GetInstance[TImplementation](ScopedRegistration registration, Scope scope)
at SimpleInjector.Advanced.Internal.LazyScopedRegistration`1.GetInstance(Scope scope)
at lambda_method(Closure )
at SimpleInjector.Scope.CreateAndCacheInstanceInternal(ScopedRegistration registration, Dictionary`2 cache)
at SimpleInjector.Scope.GetInstanceInternal(ScopedRegistration registration)
at SimpleInjector.Scope.GetInstance[TImplementation](ScopedRegistration registration, Scope scope)
at SimpleInjector.Advanced.Internal.LazyScopedRegistration`1.GetInstance(Scope scope)
at lambda_method(Closure )
at SimpleInjector.InstanceProducer.GetInstance()
````

**Comment by dotnetjunkie** [comment]
Thank you for reporting this. This is most likely a concurrency bug. I will investigate and will get back to you a.s.a.p.

**Comment by dotnetjunkie** [comment]
[bug-883](https://github.com/simpleinjector/SimpleInjector/tree/bug-883) branch created.

**Comment by dotnetjunkie** [comment]
Okay, this is a quite painful mistake. I made sure that the calls to `AsyncDisposableTypeCache.IsAsyncDisposable` where made within the context of a `lock`, not realizing that the object being locked on was an instance field of `Scope`, while the dictionary was static. This still allows multiple threads to access the dictionary concurrently as the lock is on the `Scope` instance level.

Working on a fix right now.

**Comment by dotnetjunkie** [comment]
Fixed in [v5.2.1](https://github.com/simpleinjector/SimpleInjector/releases/tag/v5.2.1)
